---
title: "AllelePurge_Fig2Table2"
author: "Audrey"
date: "30/04/2020"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Load libraries
```{r}
library(readr)
library(ggplot2)
library(tidyverse)
library(devtools)
library(gridExtra)
library(effects)
library(emmeans)
library(car)
library(lme4)
library(binom)
library(plotrix)
```

Clean data
```{r}
data <- read.csv("../data/AW_Allele_purge_counts.csv")

data_freq <- mutate(data, 
                    mut_male_freq = mutant_male/100, 
                    mut_fem_freq = mutant_fem/100)

data_freq$wt_total <- with(data_freq, wt_male + wt_fem)

data_freq$mutant_total <- with(data_freq, mutant_male + mutant_fem)

#data_freq$dummy_offset <- qlogis(0.7) # offset for logistic regression.
# the 0.7 is for allele frequencies, but for this initial analysis limited to genotypic frequencies.

data_freq <- mutate(data_freq, 
                   dummy_offset = ifelse(
  mutant_type == "yellow" | mutant_type == "white" | mutant_type == "forked", 
  qlogis(0.595), qlogis(0.49)))

#changed dummy offset so it was starting mutant genotypic frequency (0.49 for second chromosome mutations and 0.595 for X-linked mutations) instead of starting allele frequency

data_freq$replicate <- as.factor(data_freq$replicate)

data_freq <- na.omit(data_freq) #removes brown NT R2 after generation 4 (after replicate died)

```

Model we used.
Replicate as a random effect.
Note that we don't fit an intercept because this is known, but we include genotypic frequencies as an offset.

Remember, that these are not set up as treatment contrasts for the g:treatment

This code includes all replicates, including the third forked replicate. The ANOVA output is used in Table S2 and the contrast estimates generated from emtrends are used in Table S3 of the MS.
```{r}
mutants_glmm <- glmer( cbind(mutant_total, wt_total) ~
                  0 
                  + generation:treatment 
                  + generation:mutant_type
                  + generation:treatment:mutant_type
                  + (0 + generation | mutant_type:treatment:replicate) + 
                    offset(dummy_offset), 
                family = "binomial", 
                data = data_freq)

# checking whether convergence issue is worrisome.
# Nothing to worry about, estimates all really similar.
#fits_check <- allFit(mutants_glmm)
#summary(fits_check)

#summary(mutants_glmm)
Anova(mutants_glmm) 


est <- emtrends(mutants_glmm, 
                c("mutant_type", "treatment"), 
                var = "generation", transform = "none")

est
plot(est)

pairs(emtrends(mutants_glmm, 
               c("treatment"), 
               var = "generation", transform = "none"))

```

Separate model for each sex while including all replicates
The ANOVA outputs generated are used in Table S2 and the contrast estimates generated from emtrends are used in Table S3 of the MS.
```{r}
male_glmm_full <- glmer( cbind(mutant_male, wt_male) ~
                  0 
                  + generation:treatment 
                  + generation:mutant_type
                  + generation:treatment:mutant_type
                  + (0 + generation | mutant_type:treatment:replicate)
                  + offset(dummy_offset), 
                  family = "binomial", 
                  data = data_freq)

#summary(male_glmm_full)
Anova(male_glmm_full)

est_m_full <- emtrends(male_glmm_full, 
                 c("treatment","mutant_type"), 
                 var = "generation", transform = "none")

est_m_full
pairs(emtrends(male_glmm_full, 
               c("treatment"), 
               var = "generation", transform = "none"))

female_glmm_full <- glmer( cbind(mutant_fem, wt_fem) ~
                  0 
                  + generation:treatment 
                  + generation:mutant_type
                  + generation:treatment:mutant_type
                  + (0 + generation | mutant_type:treatment:replicate)
                  + offset(dummy_offset), 
                  family = "binomial", 
                  data = data_freq)

#summary(female_glmm_full)
Anova(female_glmm_full)

est_f_full <- emtrends(female_glmm_full, 
                 c("treatment","mutant_type"), 
                 var = "generation", transform = "none")

est_f_full
pairs(emtrends(female_glmm_full, 
               c("treatment"), 
               var = "generation", transform = "none"))
```


Model without third forked replicate, how does this change the model?
The ANOVA output is used in Table 1 and the contrast estimates generated from emtrends is used in Table 2 of the MS. This code is also needed to generate Figure 2 of the MS.
```{r}
forkSCT3 <- (data_freq[data_freq$mutant_type == "forked" & data_freq$replicate == "3" &
                        data_freq$treatment == "SCT",])
nofork3 <- data_freq %>% anti_join(forkSCT3)

nofork3_glmm <- glmer( cbind(mutant_total, wt_total) ~
                  0 
                  + generation:treatment 
                  + generation:mutant_type
                  + generation:treatment:mutant_type
                  + (0 + generation | mutant_type:treatment:replicate)
                  + offset(dummy_offset), 
                  family = "binomial", 
                  data = nofork3)

#summary(nofork3_glmm)
Anova(nofork3_glmm)

est2 <- emtrends(nofork3_glmm, 
                 c("treatment","mutant_type"), 
                 var = "generation", transform = "none")

est2
plot(est2)

pairs(emtrends(nofork3_glmm, 
               c("treatment"), 
               var = "generation", transform = "none"))
```

Separate sexes models
The ANOVA outputs are used in Table 1 and the contrast estimates generated from emtrends are used in Table 2 of the MS.
```{r}
male_glmm <- glmer( cbind(mutant_male, wt_male) ~
                  0 
                  + generation:treatment 
                  + generation:mutant_type
                  + generation:treatment:mutant_type
                  + (0 + generation | mutant_type:treatment:replicate)
                  + offset(dummy_offset), 
                  family = "binomial", 
                  data = nofork3)

#summary(male_glmm)
Anova(male_glmm)

est_m <- emtrends(male_glmm, 
                 c("treatment","mutant_type"), 
                 var = "generation", transform = "none")

est_m
plot(est_m)
pairs(emtrends(male_glmm, 
               c("treatment"), 
               var = "generation", transform = "none"))

est_m <- as.data.frame(est_m)


female_glmm <- glmer( cbind(mutant_fem, wt_fem) ~
                  0 
                  + generation:treatment 
                  + generation:mutant_type
                  + generation:treatment:mutant_type
                  + (0 + generation | mutant_type:treatment:replicate)
                  + offset(dummy_offset), 
                  family = "binomial", 
                  data = nofork3)

#summary(female_glmm)
Anova(female_glmm)

est_f <- emtrends(female_glmm, 
                 c("treatment","mutant_type"), 
                 var = "generation", transform = "none")

est_f
plot(est_f)
pairs(emtrends(female_glmm, 
               c("treatment"), 
               var = "generation", transform = "none"))

est_f <- as.data.frame(est_f)

```

PLOTS

Current plot in the manuscript - Figure 2
```{r}
#set up
#do separate for each allele type
predicted_glmm_auto <- as.data.frame(predictorEffects(nofork3_glmm, ~ generation, 
                                              focal.levels = 11, 
                                              fixed.predictors = 
                                                list(offset = qlogis(0.49))))
predicted_glmm_auto <- as.data.frame(predicted_glmm_auto)
predicted_glmm_auto <- subset(predicted_glmm_auto, generation.mutant_type == "brown" |
                                              generation.mutant_type == "plexus" |
                                              generation.mutant_type == "vestigial")

predicted_glmm_x <- as.data.frame(predictorEffects(nofork3_glmm, ~ generation, 
                                              focal.levels = 11, 
                                              fixed.predictors = 
                                                list(offset = qlogis(0.595))))
predicted_glmm_x <- as.data.frame(predicted_glmm_x)
predicted_glmm_x <- subset(predicted_glmm_x, generation.mutant_type == "forked" |
                                              generation.mutant_type == "white" |
                                              generation.mutant_type == "yellow")

#without the offset argument the intercept is averaged. Can set the offset to either 0.49 (autosomal) or 0.595 (x-linked) 


#get data frame in order for plotting
averaged_data_test <- nofork3 %>%
  group_by(mutant_type, treatment, generation) %>% 
  summarise(mutant_freq = mean((mut_male_freq + mut_fem_freq)/2), 
            sd = sd((mut_male_freq + mut_fem_freq)/2),
            mutant_count = mean(mutant_total))

predicted_glmm_both <- rbind(predicted_glmm_auto, predicted_glmm_x)
predicted_glmm_both <- predicted_glmm_both %>% 
  mutate(generation.mutant_type = factor(generation.mutant_type, levels = c("brown", "forked", "plexus", "vestigial", "white", "yellow"))) %>% 
  arrange(generation.mutant_type)
averaged_data_test <- data.frame(averaged_data_test, predicted_glmm_both[, 4:7])

#plot
linear_plot <- ggplot(data = averaged_data_test, aes(x = generation, y = mutant_freq, colour = treatment)) +
  theme_classic(base_size = 24) +
  theme(strip.text = element_text(face = "italic")) +
  labs(x = "Generation", y = "Mutant Genotypic Frequency") +
  geom_pointrange(aes(ymin = (mutant_freq - sd), ymax = (mutant_freq + sd)), size = 1, shape = 16, 
                  position = position_jitterdodge(dodge.width = 0.4)) +
  geom_ribbon(aes(ymin = generation.lower, ymax = generation.upper, fill = treatment), 
              alpha = 0.4) +
  facet_wrap(~mutant_type) +
  scale_colour_manual(breaks = c("NT", "UCT", "SCT"), values = c("#616161", "#E69F00", "#039BE5")) +
  scale_x_continuous(breaks = seq(0, 10, 2)) +
  scale_y_continuous(breaks = seq(0, 1, 0.3)) +
  scale_fill_manual(breaks = c("NT", "UCT", "SCT"), values = c("#616161", "#E69F00", "#039BE5")) +
  theme(legend.position = "none")

linear_plot

#contrasts
est2 <- as.data.frame(est2)
contrasts_plot <- ggplot(data = est2, aes(x = treatment, y = generation.trend)) +
  geom_pointrange(aes(ymin = asymp.LCL, ymax = asymp.UCL, colour = treatment), size = 1) +
  coord_flip() +
  facet_wrap(~mutant_type, ncol = 1, strip.position = "right") +
  scale_colour_manual(breaks = c("NT", "UCT", "SCT"), values = c("#616161", "#E69F00", "#039BE5")) +
  theme_classic(base_size = 24) +
  labs(x = NULL, y = "Generation Trend", colour = "Treatment") +
  theme(axis.text.y = element_blank(), axis.line.y = element_blank(), axis.ticks.y = element_blank())

#final plot

Fig2 <- grid.arrange(linear_plot, contrasts_plot, nrow = 1, widths = c(1.5, 1))

```

Not separating for mutations - Figure S3
```{r}
#new model
nofork3_glmm_treat <- glmer( cbind(mutant_total, wt_total) ~
                  0 
                  + generation:treatment 
                  + (0 + generation | mutant_type:treatment:replicate), 
                  family = "binomial", 
                  data = nofork3)

est3 <- emtrends(nofork3_glmm_treat, 
                 c("treatment"), 
                 var = "generation", transform = "none")

est3
plot(est3)

pairs(emtrends(nofork3_glmm_treat, 
               c("treatment"), 
               var = "generation", transform = "none"))

#set up
predicted_glmm_treat <- as.data.frame(predictorEffects(nofork3_glmm_treat, ~ generation, 
                                              focal.levels = 11,
                                              fixed.predictors = 
                                                list(offset = qlogis(0.5425))))

predicted_glmm_treat <- as.data.frame(predicted_glmm_treat) 

#get data frame in order for plotting
averaged_data_test_treat <- nofork3 %>%
  group_by(treatment, generation) %>% 
  summarise(mutant_freq = mean((mut_male_freq + mut_fem_freq)/2), 
            sd = sd((mut_male_freq + mut_fem_freq)/2),
            mutant_count = mean(mutant_total))

averaged_data_test_treat <- data.frame(averaged_data_test_treat, predicted_glmm_treat[ ,3:6])

#plot
linear_treat <- ggplot(data = averaged_data_test_treat, aes(x = generation, y = mutant_freq, colour = treatment)) +
  theme_classic(base_size = 24) +
  theme(strip.text = element_text(face = "italic")) +
  labs(x = "Generation", y = "Mutant Genotypic Frequency") +
  geom_pointrange(aes(ymin = mutant_freq - sd, ymax = mutant_freq + sd), size = 1, shape = 16, position = position_jitterdodge(dodge.width = 0.4)) +
  geom_ribbon(aes(x = generation, ymin = generation.lower, ymax = generation.upper, fill = treatment), alpha = 0.4) +
  scale_colour_manual(breaks = c("NT", "UCT", "SCT"), values = c("#616161", "#E69F00", "#039BE5")) +
  scale_x_continuous(breaks = seq(0, 10, 2)) +
  scale_y_continuous(breaks = seq(0, 1, 0.3)) +
  scale_fill_manual(breaks = c("NT", "UCT", "SCT"), values = c("#616161", "#E69F00", "#039BE5")) +
  theme(legend.position = "none")

est3 <- as.data.frame(est3)
contrasts_treat <- ggplot(data = est3, aes(x = treatment, y = generation.trend)) +
  geom_pointrange(aes(ymin = asymp.LCL, ymax = asymp.UCL, colour = treatment), size = 1) +
  coord_flip() +
  scale_colour_manual(breaks = c("NT", "UCT", "SCT"), values = c("#616161", "#E69F00", "#039BE5")) +
  theme_classic(base_size = 24) +
  labs(x = NULL, y = "Generation Trend", colour = "Treatment") +
  theme(axis.text.y = element_blank(), axis.line.y = element_blank(), axis.ticks.y = element_blank())

#final plot
FigS3 <- grid.arrange(linear_treat, contrasts_treat, nrow = 1, widths = c(1.5, 1))

```


Plot for including only forked replicates; all replicates for the forked mutation are shown and only the forked mutation is included in this figure. This line of code needs to be used along with the code in the script "AW_Allele_purge1_singlematepairings" to generate Figure S2 of the MS. See this script for further details. 
```{r}
mut_glmm_fork <- as.data.frame(predictorEffects(mutants_glmm, ~ generation, 
                                              focal.levels = 11, 
                                              fixed.predictors = 
                                                list(offset = qlogis(0.595))))
mut_glmm_fork <- as.data.frame(mut_glmm_fork)
mut_glmm_fork <- subset(mut_glmm_fork, generation.mutant_type == "forked")

averaged_data_fork <- data_freq %>%
  subset(mutant_type == "forked") %>% 
  group_by(treatment, generation) %>% 
  summarise(mutant_freq = mean((mut_male_freq + mut_fem_freq)/2), 
            sd = sd((mut_male_freq + mut_fem_freq)/2),
            mutant_count = mean(mutant_total))

averaged_data_fork <- data.frame(averaged_data_fork, mut_glmm_fork[,4:7])

fork_plot <- ggplot(data = averaged_data_fork, aes(x = generation, y = mutant_freq, colour = treatment)) +
  theme_classic(base_size = 24) +
  theme(strip.text = element_text(face = "italic")) +
  labs(x = "Generation", y = "Mutant Genotypic \n\ Frequency") +
  geom_pointrange(aes(ymin = mutant_freq - sd, ymax = mutant_freq + sd), size = 1, shape = 16, position = position_jitterdodge(dodge.width = 0.4)) +
  geom_ribbon(aes(x = generation, ymin = generation.lower, ymax = generation.upper, fill = treatment), alpha = 0.4) +
  scale_colour_manual(breaks = c("NT", "UCT", "SCT"), values = c("#616161", "#E69F00", "#039BE5")) +
  scale_x_continuous(breaks = seq(0, 10, 2)) +
  scale_y_continuous(breaks = seq(0, 1, 0.3)) +
  scale_fill_manual(breaks = c("NT", "UCT", "SCT"), values = c("#616161", "#E69F00", "#039BE5")) +
  theme(legend.position = "none")

est <- as.data.frame(est)

contrasts_fork <- ggplot(data = subset(est, mutant_type == "forked"), 
                         aes(x = treatment, y = generation.trend)) +
  geom_pointrange(aes(ymin = asymp.LCL, ymax = asymp.UCL, colour = treatment), size = 1) +
  coord_flip() +
  scale_colour_manual(breaks = c("NT", "UCT", "SCT"), values = c("#616161", "#E69F00", "#039BE5")) +
  theme_classic(base_size = 24) +
  labs(x = NULL, y = "Generation Trend", colour = "Treatment") +
  theme(axis.text.y = element_blank(), axis.line.y = element_blank(), axis.ticks.y = element_blank())

contrasts_fork

grid.arrange(fork_plot, contrasts_fork, nrow = 1, widths = c(1.5, 1))
```


---- ADDITIONAL ANALYSES; Not in MS

Plot like Chris'
Cannot easily generate confidence intervals and standard errors for mixed effect models.
Looks like Chris used just the fixed effects (glm) to generate figure (or at least the shaded area)
Code still works if using the mixed model but the random effects are ignored so the confidence bands generated are the same
Keeping this chunk as a glm so it is not confused that ONLY fixed effects are used

```{r}
nofork3_glm <- glm( cbind(mutant_total, wt_total) ~
                  0 
                  + generation:treatment 
                  + generation:mutant_type
                  + generation:treatment:mutant_type
                  + offset(dummy_offset), 
                  family = "binomial", 
                  data = nofork3)

predicted <- data.frame(generation = rep(seq(from = 0, to = 10, by = 1), 18),
                        mutant_type = c(rep("brown", 33), rep("forked", 33), rep("plexus", 33),
                                         rep("vestigial", 33), rep("white", 33), rep("yellow", 33)),
                        treatment = rep(c(rep("NT", 11), rep("SCT", 11), rep("UCT", 11)), 6))
#fix this because brown has some extra replicates and forked has one less?
predicted <- mutate(predicted, dummy_offset = 
                      ifelse(mutant_type == "yellow" | mutant_type == "white" | mutant_type == "forked",
                             qlogis(0.595), qlogis(0.49)))

y.pred <- data.frame(predicted, predict(nofork3_glm, newdata = predicted, type = "response", 
                                        se.fit = T))
y.pred$y.pred.up <- y.pred$fit + 2*y.pred$se
y.pred$y.pred.low <- y.pred$fit - 2*y.pred$se

averaged_data <- nofork3 %>%
  group_by(mutant_type, treatment, generation) %>% 
  summarise(mutant_freq = mean((mut_male_freq + mut_fem_freq)/2), 
            sd = sd((mut_male_freq + mut_fem_freq)/2),
            mutant_count = mean(mutant_total))
averaged_data <- data.frame(averaged_data, y.pred[ ,4:9])


#Add confidence intervals for points

#library(Rmisc)
#avg_data <- group.CI(mutant_freq ~ mutant_type + generation + treatment,
                     #data = nofork3,
                     #ci = 0.95) 
#confidence intervals go past 0 and 1 which does not make sense for this data
#colnames(avg_data)[5] <- "mutant_freq"

data_confs <- binom.confint(x = averaged_data$mutant_count, n = 200, conf.level = 0.95, method = "logit")
#is logit the right method to use?

averaged_data <- data.frame(averaged_data, data_confs)

  
linear_plot2 <- ggplot(data = averaged_data, aes(x = generation, y = mutant_freq, colour = treatment)) +
  geom_point(size = 4, shape = 16) +
  facet_wrap(~mutant_type) +
  theme_classic(base_size = 24) +
  theme(strip.text = element_text(face = "italic")) +
  labs(x = "Generation", y = "Mutant Frequency") +
  geom_linerange(aes(ymin = lower, ymax = upper)) + 
  geom_ribbon(aes(x = generation, ymin = y.pred.low, ymax = y.pred.up, fill = treatment), alpha = 0.5) +
  scale_colour_manual(breaks = c("NT", "UCT", "SCT"), values = c("#616161", "#E69F00", "#039BE5")) +
  scale_x_continuous(breaks = seq(0, 10, 2)) +
  scale_y_continuous(breaks = seq(0, 1, 0.3)) +
  scale_fill_manual(breaks = c("NT", "UCT", "SCT"), values = c("#616161", "#E69F00", "#039BE5")) +
  theme(legend.position = "none")

#The points are the average mutant frequencies of each mutant type over time with the lines representing the 95% confidence intervals. The ribbons are estimated confidence intervals based on the models' fixed effects.
#Need to check if code is correct. 

grid.arrange(linear_plot2, contrasts_plot, nrow = 1, widths = c(1.5, 1))
```

Comparing the sexes
```{r}
nofork3 <- nofork3 %>% mutate(model_male = predict(male_glmm, type = "response"), 
                              model_fem = predict(female_glmm, type = "response"))

ggplot(data = nofork3, aes(generation, mut_male_freq, colour = treatment)) +
  geom_point(aes(generation, mut_fem_freq, colour = treatment), 
             size = 4, alpha = 0.4, shape = 17) +
  geom_point(aes(generation, mut_male_freq, colour = treatment),
             size = 4, alpha = 0.4, shape = 16) +
  facet_wrap(~mutant_type) +
  theme_classic(base_size = 24) +
  theme(strip.text = element_text(face = "italic")) +
  labs(x = "Generation", y = "Mutant Frequency") +
  scale_color_manual(breaks = c("NT", "UCT", "SCT"), values = c("#616161", "#E69F00", "#039BE5")) +
  scale_x_continuous(breaks = seq(0, 10, 2)) +
  scale_y_continuous(breaks = seq(0, 1, 0.3)) +
  geom_smooth(aes(x = generation, y = model_male, colour = treatment), size = 2, se = FALSE) +
  geom_smooth(aes(x = generation, y = model_fem, colour = treatment), size = 2, linetype = 2, se = FALSE)
```
